src/main/java/com/snowflake/kafka/connector/internal/SnowflakeConnectionService.java

```diff
@@ -91,6 +91,41 @@ public interface SnowflakeConnectionService {
    */
   boolean isTableCompatible(String tableName);
 
+  /**
+   * Check whether the user has the role privilege to do schema evolution and whether the schema
+   * evolution option is enabled on the table
+   *
+   * @param tableName the name of the table
+   * @param role the role of the user
+   * @return whether table and role has the required permission to perform schema evolution
+   */
+  boolean hasSchemaEvolutionPermissionForRole(String tableName, String role);
+
+  /**
+   * Check if the role used by the connector has the permission to do schema evolution on the table
+   * through a cached map
+   *
+   * @param tableName the name of the table
+   * @return whether we have the permission to do schema evolution on the table
+   */
+  boolean hasSchemaEvolutionPermission(String tableName);
```

Comment:
Can we combine this two APIs together?

Reply:
We can get rid of this function and can just use above one but pass in the default role?

Reply:
So far the `hasSchemaEvolutionPermission(String tableName)` reads the cache. If we let it call the other with the default role, we will be continuously querying the permission with each buffer to be inserted.